### PR TITLE
Allow empty reference summaries

### PR DESCRIPTION
**Pull request by JohnGiorgi**
# What does this PR do?

The `run_summarization.py` has a check to skip examples where either the `text_column` or `summary_column` is `None`.  However, the way the check was written would catch any falsely values, like an empty string.

This caused all examples to be skipped for datasets where either the `text_column` or `summary_column` was the empty string (e.g. the test set of the [MS^2 dataset](https://huggingface.co/datasets/allenai/mslr2022)).

This PR just updates the check so it looks for `None` values explicitly.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by LysandreJik** [comment]
@ydshieh, would you like to take a look at what is proposed here?

**Comment by ydshieh** [comment]
@JohnGiorgi I am not familiar with this dataset. But I am wondering if a model trained on this dataset is expected to learn to deal with empty document (this seems to strange) and/or learn to predict empty summary for some document?

**Comment by JohnGiorgi** [comment]
@ydshieh They are just using the empty string as a placeholder. The reference summaries are held-out and not available publically. I imagine there could be other cases like this

**Comment by ydshieh** [comment]
Log a warning sounds good to me. But before doing this, I am wondering what would be the benefits to process those held-out (empty string) test examples. Those examples won't be useful, right?

**Comment by JohnGiorgi** [comment]
You would still want to generate predicts for example even if they don’t have reference summaries

**Comment by JohnGiorgi** [comment]
@ydshieh Okay hows that? The only change now is that examples with empty reference summaries are still included, but a warning will be logged.

**Comment by ydshieh** [comment]
Thank you @JohnGiorgi, this LGTM overall!

The logging on each example with empty summary might be too spam. (Imagine the test dataset has 10K examples).
We can probably set a flag. If an empty summary is found -> and if flag is False -> warning -> set flag to True.

Let's wait one of the core maintainers to give a final review.

**Comment by JohnGiorgi** [comment]
> Thank you @JohnGiorgi, this LGTM overall!
> 
> The logging on each example with empty summary might be too spam. (Imagine the test dataset has 10K examples). We can probably set a flag. If an empty summary is found -> and if flag is False -> warning -> set flag to True.
> 
> Let's wait one of the core maintainers to give a final review.

Ah very good point! Is there an example somewhere else in the codebase of something like this? I'm happy to update my change following that

**Comment by ydshieh** [comment]
Hi! Something like `deprecation_warnings`

https://github.com/huggingface/transformers/blob/30992ef0d911bdeca425969d210771118a5cd1ac/src/transformers/tokenization_utils_base.py#L1520

could work, but in the training script, I think a single Boolean variable is enough.

**Comment by JohnGiorgi** [comment]
Okay I will just keep it simple! How is this?

File: examples/pytorch/summarization/run_summarization.py [hunk 1]
```diff
@@ -512,11 +512,15 @@ def main():
...
+        warned_about_empty = False
         inputs, targets = [], []
         for i in range(len(examples[text_column])):
-            if examples[text_column][i] and examples[summary_column][i]:
+            if examples[text_column][i]:
+                if not examples[summary_column][i] and not warned_about_empty:
+                    logger.warning(
+                        "Found an empty reference summary. This will be included in the preprocessed dataset."
+                    )
+                    warned_about_empty = True
```
Comment (JohnGiorgi) [inline examples/pytorch/summarization/run_summarization.py:523]: Okay I will just keep it simple, hows this?

Comment (ydshieh) [reply examples/pytorch/summarization/run_summarization.py:523]: I think `warned_about_empty` should be defined outside `preprocess_function`, as it might get called several times (depending on the processing batch size).

Comment (JohnGiorgi) [reply examples/pytorch/summarization/run_summarization.py]: Good catch, done.

**Comment by JohnGiorgi** [comment]
Hi @ydshieh, is there anything blocking this that I can address?

**Comment by JohnGiorgi** [comment]
Ah, didn't look carefully at why the build was failing. Thanks, @ydshieh, your solution causes all the tests to pass!

**Review by ydshieh — APPROVED** [review]
Thanks, @JohnGiorgi!

@sgugger Could you also take a look? This PR is to allow examples with empty summary for prediction purpose.

I think it is safe to assume the training and evaluation datasets won't have examples with empty summary. But if you think it's necessary, we can also try to make the `preprocess_function` behave differently in 2 modes.

File: examples/pytorch/summarization/run_summarization.py [hunk 2]
```diff
@@ -512,11 +512,14 @@ def main():
         )
 
     def preprocess_function(examples):
-        # remove pairs where at least one record is None
-
         inputs, targets = [], []
         for i in range(len(examples[text_column])):
-            if examples[text_column][i] and examples[summary_column][i]:
+            if examples[text_column][i]:
+                if not examples[summary_column][i] and not empty_summary_warning["warned"]:
```
Comment (ydshieh) [inline examples/pytorch/summarization/run_summarization.py:518]: Not very sure if we could get `None` value for `examples[summary_column][i]`.

```
if examples[text_column][i] and examples[summary_column][i] is not None:
```

**Comment by sgugger** [comment]
This is too complicated for one of the example, which are just examples, not production-ready apps that should work in every case. Readability is more important than fixing this edge case IMO.

**Comment by ydshieh** [comment]
> not production-ready apps that should work in every case. Readability is more important than fixing this edge case IMO.

Fair point.

@JohnGiorgi Still thank you a lot for the PR. I am sorry that I haven't thought in another angle as a repository maintainer, and let you spend quite some time working on several of my suggestions.

You can definitely tweak the example script to meet your own needs. I am closing the issue.
